test/hotspot/jtreg/compiler/lib/template_framework/TemplateFrame.java

```diff
@@ -78,22 +131,43 @@ void addHashtagReplacement(String key, String value) {
         if (!Renderer.isValidHashtagOrDollarName(key)) {
             throw new RendererException("Is not a valid hashtag replacement name: '" + key + "'.");
         }
-        if (hashtagReplacements.putIfAbsent(key, value) != null) {
-            throw new RendererException("Duplicate hashtag replacement for #" + key);
+        String previous = findHashtagReplacementInScopes(key);
+        if (previous != null) {
+            throw new RendererException("Duplicate hashtag replacement for #" + key + ". " +
+                                        "previous: " + previous + ", new: " + value);
+        }
+        if (isTransparentForHashtag) {
+            parent.addHashtagReplacement(key, value);
+        } else {
+            hashtagReplacements.put(key, value);
         }
     }
 
     String getHashtagReplacement(String key) {
         if (!Renderer.isValidHashtagOrDollarName(key)) {
             throw new RendererException("Is not a valid hashtag replacement name: '" + key + "'.");
         }
+        String value = findHashtagReplacementInScopes(key);
+        if (value != null) {
+            return value;
+        }
+        throw new RendererException("Missing hashtag replacement for #" + key);
+    }
+
+    private String findHashtagReplacementInScopes(String key) {
         if (hashtagReplacements.containsKey(key)) {
             return hashtagReplacements.get(key);
         }
-        throw new RendererException("Missing hashtag replacement for #" + key);
+        if (!isInnerScope) {
+            return null;
+        }
+        return parent.findHashtagReplacementInScopes(key);
     }
 
     void setFuelCost(float fuelCost) {
         this.fuelCost = fuelCost;
+        if (this.isTransparentForFuel) {
+            this.parent.setFuelCost(fuelCost);
+        }
```

Comment:
`this` can be omitted:

```suggestion
        if (isTransparentForFuel) {
            parent.setFuelCost(fuelCost);
        }
```